### PR TITLE
Add FireEye ETP Event Collector

### PR DESCRIPTION
**Pull request by JasBeilin**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-5922

## Description
Adding new event collector to allow fetch of alerts, email-trace and activity-log from users.

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    name: proxy
+    type: 8
+    section: Connect
+    advanced: true
+  - display: Fetch outbound traffic
+    name: outbound_traffic
+    type: 8
+    section: Collect
+    advanced: true
+    additionalinfo: Outbound traffic will be fetched in addition to inbound traffic
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector.yml:54]:
```suggestion
    additionalinfo: Outbound traffic will be fetched in addition to inbound traffic.
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  - display: Hide sensitive details from email
+    name: hide_sensitive
+    type: 8
+    section: Collect
+    advanced: true
+    defaultvalue: 'true'
+    additionalinfo: Hide subject and attachments details from emails
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector.yml:61]:
```suggestion
    additionalinfo: Hide subject and attachments details from emails.
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+FireEye Email Threat Prevention Event Collector allows you to fetch events of type Alert, Email Trace ad Activity Log from your system.
+
+# Configuring API keys
+Follow these steps to configure API keys:
+1. Log in to the Email Security — Cloud Web Portal or IAM console.
+2. Click on My Settings in the top navigation bar.
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md:6]:
```suggestion
2. Click **My Settings** in the top navigation bar.
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+3. Click the API Keys tab in the IAM console.
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md:7]:
```suggestion
3. Click the **API Keys** tab in the IAM console.
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+4. Click Create API Key.
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md:8]:
```suggestion
4. Click **Create API Key**.
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+5. On the Manage API Key page, specify the following:
+   - API key name.
+   - Expiration time for the API key. The expiration time of API keys should be set as “100d” for 100 days, or “1y” for 1 year, for example.
+   - Products. Select both “Email Threat Prevention” and “Identity Access Management”.
+6. Select all entitlements as shown below.
+7. To download or copy an API key, click the download or copy icon in the bottom right corner.
+8. Click Create API Key.
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md:15]:
```suggestion
8. Click **Create API Key**.
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+   
+# Permissions
+For any API access, the following entitlements are required:
+- iam.users.browse
+- iam.orgs.self.read
+
+For accessing alerts APIs, the following additional entitlements are required:
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md:22]:
```suggestion
For accessing alerts APIs, the following additional entitlement is required:
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+- etp.alerts.read
+
+For accessing trace APIs, the following additional entitlements are required:
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md:25]:
```suggestion
For accessing trace APIs, the following additional entitlement is required:
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+## Cloud service regions
+Use the URLs for the region that hosts your Email Security — Cloud service:
+
+- US Instance: https://etp.us.fireeye.com/
+- EMEA Instance: https://etp.eu.fireeye.com
+- APJ Instance: https://etp.ap.fireeye.com/
+- USGOV Instance: https://etp.us.fireeyegov.com/
+- CA Instance: https://etp.ca.fireeye.com/
+
+For more information, check [Official Product Documentation](https://docs.trellix.com/bundle/etp_api/page/UUID-98fd1a2c-382d-130b-00c5-b9be402fe660_1.html#idm44910884288000).
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector_description.md:37]:
```suggestion
For more information, see the [Official Product Documentation](https://docs.trellix.com/bundle/etp_api/page/UUID-98fd1a2c-382d-130b-00c5-b9be402fe660_1.html#idm44910884288000).
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/README.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    | **Parameter** | **Description** | **Required** |
+    | --- | --- | --- |
+    | Server URL (e.g., https://etp.us.fireeye.com) |  | True |
+    | API Secret Key | The API Key allows you to integrate with the FireEye ETP. | True |
+    | Maximum number of Alerts to fetch. | The maximum number of Alert events to fetch from FireEye ETP. |  |
+    | Maximum number of Email Trace to fetch. | The maximum number of Email Trace events to fetch from FireEye ETP. |  |
+    | Maximum number of Activity Log fetch. | The maximum number of Activity Log events to fetch from FireEye ETP. |  |
+    | Trust any certificate (not secure) |  |  |
+    | Use system proxy settings |  |  |
+    | Fetch outbound traffic | Outbound traffic will be fetched in addition to inbound traffic |  |
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/README.md:18]:
```suggestion
    | Fetch outbound traffic | Outbound traffic will be fetched in addition to inbound traffic. |  |
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/README.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    | Hide sensitive details from email | Hide subject and attachments details from emails |  |
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/README.md:19]:
```suggestion
    | Hide sensitive details from email | Hide subject and attachments details from emails. |  |
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/README.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
+Use this integration to fetch email security incidents from FireEye ETP as XSIAM events.
+
+## Configure FireEye ETP Event Collector on Cortex XSOAR
+
+1. Navigate to **Settings** > **Integrations** > **Servers & Services**.
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/README.md:5]:
```suggestion
1. Navigate to **Settings** > **Configurations** > **Data Collection** > **Automation & Feed Integrations**.
```

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/README.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+60 requests to Trace APIs (parallel or sequential)
+60 requests to Alert APIs (parallel or sequential)
+60 requests to Quarantine APIs (parallel or sequential)
+
+Within the minute, the 61st request to any of these APIs would throw a rate limit exceeded error.
+
+The rate limit applies to the customer as a whole. This means that if the customer has multiple admin users who have generated API Keys, the rate limit is applicable at the customer level and not per API key. 
+
+## Commands
+You can execute these commands from the Cortex XSOAR CLI, as part of an automation, or in a playbook.
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/README.md:39]:
```suggestion
You can execute these commands from the Cortex XSIAM CLI, as part of an automation, or in a playbook.
```

File: Packs/FireEyeETP/ReleaseNotes/1_3_0.md [hunk 14]
```diff
@@ -0,0 +1,11 @@
+#### Integrations
+
+##### New: FireEye ETP Event Collector
+
+- New: Use this integration to fetch email security incidents from FireEye ETP as XSIAM events.
```
Comment (ShirleyDenkberg) [inline Packs/FireEyeETP/ReleaseNotes/1_3_0.md:5]:
```suggestion
New: Use this integration to fetch email security incidents from FireEye ETP as XSIAM events.
```

**Comment by ShirleyDenkberg** [comment]
@yuvalbenshalom Doc review completed.

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
looks good. see my comments.

general comment, please check if passing **args is really needed. in general I prefer to avoid it.

File: Packs/FireEyeETP/pack_metadata.json [hunk 15]
```diff
@@ -2,7 +2,7 @@
     "name": "FireEye ETP",
     "description": "FireEye Email Threat Prevention (ETP Cloud) is a cloud-based platform that protects against advanced email attacks.",
     "support": "xsoar",
-    "currentVersion": "1.2.4",
```
[the remark this answers is not in the document]
Comment (JasBeilin) [reply Packs/FireEyeETP/pack_metadata.json:5]: Yes, by design.

File: Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector.py [hunk 16]
```diff
@@ -0,0 +1,541 @@
...
+        return events, CommandResults(readable_output=hr)
+
+    def fetch_alerts(self, event_type: EventType, start_time: datetime,
+                     end_time: Optional[datetime], fetched_ids: set = set(), **args) -> tuple[list[dict], datetime]:
+        res_count = 0
+        res: list[dict] = []
+        results_left = True
+        iso_start_time = start_time.isoformat()[:-3]  # formating to 3 digit's microseconds
+        #  Running as long as we have not reached the amount of event or the time frame requested.
+        while results_left and res_count < event_type.client_max_fetch and (not (end_time and end_time <= start_time)):
```
[the remark this answers is not in the document]
Comment (JasBeilin) [reply Packs/FireEyeETP/Integrations/FireEyeETPEventCollector/FireEyeETPEventCollector.py:232]: It is due to the difference in paging param, and functions (for example if ID is not an existing field), I thought generic will be harder to read in this case, and might be more expensive in running time.

**Comment by JasBeilin** [comment]
Some changes post Demo-
1. API Key display name change
2. Permission in Description are now joined
3. Handling a wrong credential error.
4. mapping changes
5. sensitive data will not be deleted but replaced with "hidden data"
6. better explanation of the outbound param affects
7. better get-command warning of use.

**Review by yuvalbenshalom — APPROVED** [review]

**Pull request merged**